opm/simulators/wells/GasLiftSingleWellGeneric.cpp

```diff
@@ -748,6 +761,11 @@ runOptimizeLoop_(bool increase)
             state.stop_iteration = true;
         }
     }
+    // we only iterate if well is under thp control
+    if (!state.checkThpControl()) {
+        state.stop_iteration = true;
+    }
```

Comment:
I am probably missing something fundamental, but it is not clear to me why THP control is checked this late? For example, at line 719 at the beginning of this function, `runOptimizeLoop()_`, we call  `computeInitialWellRates_()` which will calculate the well rates given the current ALQ at the THP limit. The ALQ is also adjusted at 734 by the call to `maybeAdjustALQbeforeOptimizeLoop_()` which also use the THP limit pressure to calculate the well rates. Does it make sense to use the THP limit pressure like this when the well is not under THP control? Shouldn't this check be at the beginning of this function? Or even earlier, for example in `gasLiftOptimizationStage1()` in `BlackoilWellModel_impl.hpp` ([line 940](https://github.com/OPM/opm-simulators/blob/master/opm/simulators/wells/BlackoilWellModel_impl.hpp#L940)) : 
```
                    if (group_info.hasWell(well->name())) {
```
The `hasWell()` method of the group info object could pick out the wells with THP control. In fact it already has a check for THP control, see `GasLiftGroupInfo.cpp` [line 280](https://github.com/OPM/opm-simulators/blob/master/opm/simulators/wells/GasLiftGroupInfo.cpp#L280) : 
```
   if (this->optimize_only_thp_wells_) {
```

Reply:
If I remember correctly the thp check was moved down the stack to allow for wells under group control to reduce its gaslift. A typical case for this could be that a new well opens up. Then gaslift can be reduced while still keeping the group target.